qiskit/visualization/state_visualization.py

```diff
@@ -1149,10 +1156,111 @@ def state_to_latex(state, dims=None, **args):
+    else:
+        latex_str = array_to_latex(state._data, source=True, **args)
     return prefix + latex_str + suffix
 
 
+def num_to_latex_ket(raw_value: complex, first_term: bool) -> Optional[str]:
+    """Convert a complex number to latex code suitable for a ket expression
+
+    Args:
+        raw_value: Value to convert
+        first_term: If True then generate latex code for the first term in an expression
+    Returns:
+        String with latex code or None if no term is required
+    """
+    import sympy  # runtime import
+
+    raw_value = complex(np.real_if_close(raw_value))
+    raw_value = -1j * np.real_if_close(raw_value * 1j)
+
+    value = sympy.nsimplify(raw_value, constants=(sympy.pi,), rational=False)
+    real_value = float(sympy.re(value))
+    imag_value = float(sympy.im(value))
+
+    element = ""
+    if np.abs(value) > 0:
+        latex_element = sympy.latex(value, full_prec=False)
+        two_term = real_value != 0 and imag_value != 0
+        if isinstance(value, sympy.core.Add):
+            # can happen for expressions like 1 + sqrt(2)
+            two_term = True
+        if two_term:
+            if first_term:
+                element = f"({latex_element})"
+            else:
+                element = f"+ ({latex_element})"
+        else:
+            if first_term:
+                if np.isreal(complex(value)) and value > 0:
+                    element = latex_element
+                else:
+                    element = latex_element
+                if element == "1":
+                    element = ""
+                elif element == "-1":
+                    element = "-"
+            else:
+
+                if imag_value == 0 and real_value > 0:
+                    element = "+" + latex_element
+                elif real_value == 0 and imag_value > 0:
+                    element = "+" + latex_element
+                else:
+                    element = latex_element
+                if element == "+1":
+                    element = "+"
+                elif element == "-1":
+                    element = "-"
+
+        return element
+    else:
+        return None
+
+
+def numbers_to_latex_terms(numbers: List[complex]) -> List[str]:
+    """Convert a list of numbers to latex formatted terms
+
+    The first non-zero term is treated differently. For this term a leading + is suppressed.
+
+    Args:
+        numbers: List of numbers to format
+    Returns:
+        List of formatted terms
+    """
+    first_term = True
+    terms = []
+    for number in numbers:
+        term = num_to_latex_ket(number, first_term)
+        if term is not None:
+            first_term = False
+        terms.append(term)
+    return terms
+
+
+def _state_to_latex_ket(data: List[complex]) -> str:
+    """Convert state vector to latex representation"""
+    num = int(np.log2(len(data)))
+    ket_names = [bin(i)[2:].zfill(num) for i in range(2 ** num)]
+
+    latex_terms = numbers_to_latex_terms(data)
```

Comment:
What happens when things get large?  e.g. `2**27` elements?  What should the output look like in this case?